ramips: add support for Qding QC202

rbqvq opened the pull request:
This is a smart door lock device equipped with OpenWrt 14.07 OEM
modified version Qdwrt

The OEM has closed down, This commit is intended to maximize the
remaining value of these devices. It can flash OpenWrt to become
an AP

Specification:
- SoC: MediaTek MT7628NN
- Flash: 8 MB
- RAM: 64 MB
- Power: DC 5V - 25V
- Ethernet: 1 x RJ45 (10/100 Mbps)
- Wireless radio: 802.11n 2.4g-only
- On-Board LED:
  Status 1: GPIO/43 active-low
  Status 2: GPIO/44 active-low
  Power: AlwaysOn
- Button:
  WPS / RESET: GPIO/14  active-low
- Bluetooth: CC2541 via UART1 (ttyS1) and GPIO/26-29
- RFID: MF RC522 on I2C@28
- RTC: DS1339 on I2C@68
- Shell panel (via CON1 cable)
  - LED (Swipe card area):
        Green  GPIO/3  active-high
        Red     GPIO/11 active-high
  - Matrix keypad: (active-low)
 
|Cols⟍Rows| GPIO/20 | GPIO/21 | GPIO/19 |
|---------|---------|---------|---------|
|GPIO/24|    1     |   2   |      3      |
|GPIO/25|    4     |   5   |      6      |
|GPIO/22|    7     |   8   |      9      |
|GPIO/23| BACK |   0   | ENTER   |

- UART: 1 x UART on PCB - 57600 8N1
- GPIO Relay: GPIO/42 active-high
- GPIO Buzzer: GPIO/15 active-high

Warning:
The original firmware does not use the device tree.
This device tree is written based on the content of /sys/devices/platform
and has been tested

Note:
- On the device, matrix keypad rows actually are columns, and the columns actually are rows
- The key code of the CLEAR key of the matrix keypad is BACK in the original firmware.

Issue:
- No drivers in mainline kernel for RFID and Bluetooth.

Flash Instruction:
Using SSH/Telnet:
1. Connect the board to the computer via RJ45 Ethernet
2. Login 10.10.10.1 with root password "szqdingnet123" (SSH Port 22, Telnet Port 9900)
3. Download openwrt firmware on the computer.
4. Setup a http server on computer. And use wget download openwrt firmware from computer
5. Use command "mtd -r write openwrt-ramips-mt76x8-qding_qc202-squashfs-sysupgrade.bin firmware"
   to flash

Using U-Boot WebUI:
1. Configure PC with a static IP address 10.10.10.2/24.
2. Open http://10.10.10.1/
3. Use "mkqdimg -B qc202 -f openwrt-ramips-mt76x8-qding_qc202-squashfs-sysupgrade.bin" to
   make image.
4. Upload factory.bin via U-Boot WebUI.

Original Firmware Dump / More details:
https://blog.gov.cooking/archives/research-qianding-smart-locker-and-flash.html

Original U-Boot firmware image tools:
https://gitlab.com/CoiaPrant/mkqdimg

---

Follow #17226 
I renamed branch and its auto closed.

rbqvq committed:
ramips: add support for Qding QC202

This is a smart door lock device equipped with OpenWrt 14.07 OEM
modified version Qdwrt

The OEM has closed down, This commit is intended to maximize the
remaining value of these devices. It can flash OpenWrt to become
an AP

Specification:
- SoC: MediaTek MT7628NN
- Flash: 8 MB
- RAM: 64 MB
- Power: DC 5V - 25V
- Ethernet: 1 x RJ45 (10/100 Mbps)
- Wireless radio: 802.11n 2.4g-only
- On-Board LED:
  Status 1: GPIO/43 active-low
  Status 2: GPIO/44 active-low
  Power: AlwaysOn
- Button:
  WPS / RESET: GPIO/14  active-low
- Bluetooth: CC2541 via UART1 (ttyS1) and GPIO/26-29
- RFID: MF RC522 on I2C@28
- RTC: DS1339 on I2C@68
- Shell (via CON1 cable)
  - LED (Swipe card area):
     -  Green  GPIO/3  active-high
     -  Red    GPIO/11 active-high
  - Matrix keypad: (active-low)
                 GPIO/20 GPIO/21 GPIO/19 (Rows)
        GPIO/24        1       2       3
        GPIO/25        4       5       6
        GPIO/22        7       8       9
        GPIO/23     BACK       0   ENTER
        (Cols)
- UART: 1 x UART on PCB - 57600 8N1
- GPIO Relay: GPIO/42 active-high
- GPIO Buzzer: GPIO/15 active-high

Warning:
The original firmware does not use the device tree.
This device tree is written based on the content of /sys/devices/platform
and has been tested

Note:
- On the device, matrix keypad rows actually are columns, and the columns actually are rows
- The key code of the CLEAR key of the matrix keypad is BACK in the original firmware.

Issue:
- No drivers in mainline kernel for RFID and Bluetooth.

Flash Instruction:
Using SSH/Telnet:
1. Connect the board to the computer via RJ45 Ethernet
2. Login 10.10.10.1 with root password "szqdingnet123" (SSH Port 22, Telnet Port 9900)
3. Download openwrt firmware on the computer.
4. Setup a http server on computer. And use wget download openwrt firmware from computer
5. Use command "mtd -r write openwrt-ramips-mt76x8-qding_qc202-squashfs-sysupgrade.bin firmware"
   to flash

Using U-Boot WebUI:
1. Configure PC with a static IP address 10.10.10.2/24.
2. Open http://10.10.10.1
3. Use "mkqdimg -B qc202 -f openwrt-ramips-mt76x8-qding_qc202-squashfs-sysupgrade.bin" to
   make image.
4. Upload factory.bin via U-Boot WebUI.

Original Firmware Dump / More details:
https://blog.gov.cooking/archives/research-qianding-smart-locker-and-flash.html

Original U-Boot firmware image tools:
https://gitlab.com/CoiaPrant/mkqdimg

Signed-off-by: Coia Prant <coiaprant@gmail.com>
Tested-by: Coia Prant <coiaprant@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17471
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

rbqvq commented:
mkqdimg has been rewritten in C
See https://github.com/openwrt/firmware-utils/pull/38

DragonBluep commented on target/linux/ramips/dts/mt7628an_qding_qc202.dts:
```diff
@@ -0,0 +1,175 @@
+
+&state_default {
+	gpio {
+		groups = "gpio";
```
Please add all necessary GPIO groups.

rbqvq replied:
I'm not understand here. Original firmware not use device-tree.
I think I need see more document about it.

rbqvq replied:
Fixed

rbqvq commented on target/linux/ramips/dts/mt7628an_qding_qc202.dts:
```diff
@@ -0,0 +1,175 @@
+					};
+
+					macaddr_factory_4: macaddr@4 {
+						reg = <0x4 0x6>;
```
In fact on original firmware, this mac address used for ethernet port.
wmac mac is it -1

I dont know how to setup it.

But we can confirm this mac address is unique.

rbqvq commented on target/linux/ramips/dts/mt7628an_qding_qc202.dts:
```diff
@@ -0,0 +1,175 @@
+	spidev@1{
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "linux,spidev";
```
It copied from `mt7628an_mediatek_linkit-smart-7688.dts`, I just refer to it.

Rondom commented:
Is the CC2541 maybe attached via UART? Would be maybe interesting to run custom firmware on that one.

rbqvq commented:
> Is the CC2541 maybe attached via UART? Would be maybe interesting to run custom firmware on that one.

I dont know. On original firmware, it has a kmod to read firmware from /lib/firmware then flash to CC2541.

I use IDA decompile it, it seems using GPIO to control with a fixed memory region.

rbqvq commented:
> Is the CC2541 maybe attached via UART? Would be maybe interesting to run custom firmware on that one.

Yes, You are right. I checked original firmware.
The gate access control program communicate CC2541 BLE via UART (ttyS0 / uart1).

I will update this PR to enable UART 1.

rbqvq committed:
kernel: add kmod-input-matrix-keypad package

This module contains driver for GPIO driven matrix keypad.

Signed-off-by: Coia Prant <coiaprant@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17471
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

rbqvq commented:
I'm very excited that I have completed simple reverse engineering. Here is the latest progress

- Bluetooth: CC2541 via UART1 (ttyS1) and GPIO/26-29
- Enclosure panel (via CON1 cable)
  - Matrix keypad: (active-low)
 
|Cols⟍Rows| GPIO/20 | GPIO/21 | GPIO/19 |
|---------|---------|---------|---------|
|GPIO/24|    1     |   2   |      3      |
|GPIO/25|    4     |   5   |      6      |
|GPIO/22|    7     |   8   |      9      |
|GPIO/23| BACK |   0   | ENTER   |

> The manufacturer's engineers can't seem to distinguish between rows and columns. It is not known whether this is intentional or a bug.

Note:
- On the device, matrix keypad rows actually are columns, and the columns actually are rows
- The key code of the CLEAR key of the matrix keypad is BACK in the original firmware.

rbqvq commented:
ping @hauke @dangowrt

hauke commented on target/linux/ramips/mt76x8/target.mk:
```diff
@@ -4,7 +4,7 @@
 SUBTARGET:=mt76x8
 BOARDNAME:=MT76x8 based boards
-FEATURES+=usb ramdisk small_flash
+FEATURES+=usb ramdisk rtc small_flash
```
I would prefer if we do not activate rtc support in this target. Then the RTC chip on your board will not work. 

This will increase the kernel binary for all devices in this target. Some boards do not have much flash.

rbqvq commented:
@hauke RTC support is removed.

I think this device should be merged as soon as possible and backport to 24.10.
Time is running out for 8M Flash devices.

rbqvq commented:
ping @hauke

hauke commented:
Merged it, I overlooked the first commit.
